pkg/virt-handler/cgroup/cgroup_v1_manager.go

```diff
@@ -0,0 +1,111 @@
+package cgroup
+
+import (
+	"bytes"
+	"fmt"
+	"path/filepath"
+
+	"kubevirt.io/client-go/log"
+
+	cgroup_devices "github.com/opencontainers/runc/libcontainer/cgroups/devices"
+	"github.com/opencontainers/runc/libcontainer/devices"
+
+	runc_cgroups "github.com/opencontainers/runc/libcontainer/cgroups"
+	runc_fs "github.com/opencontainers/runc/libcontainer/cgroups/fs"
+	runc_configs "github.com/opencontainers/runc/libcontainer/configs"
+)
+
+type v1Manager struct {
+	runc_cgroups.Manager
+}
+
+func newV1Manager(config *runc_configs.Cgroup, controllerPaths map[string]string, rootless bool) (Manager, error) {
+	runcManager := runc_fs.NewManager(config, controllerPaths, rootless)
+	manager := v1Manager{
+		runcManager,
+	}
+
+	return &manager, nil
+}
+
+func (v *v1Manager) GetBasePathToHostSubsystem(subsystem string) (string, error) {
+	if v.Path(subsystem) == "" {
+		return "", fmt.Errorf("controller %s does not exist", subsystem)
+	}
+	return filepath.Join(HostCgroupBasePath, subsystem), nil
+}
+
+func (v *v1Manager) Set(r *runc_configs.Resources) error {
+	// We want to keep given resources untouched
+	resourcesToSet := *r
+
+	// Adding current rules, see addCurrentRules's documentation for more info
+	requestedAndCurrentRules, err := v.addCurrentRules(r.Devices)
+	if err != nil {
+		return err
+	}
+
+	log.Log.V(loggingVerbosity).Infof("Adding current rules to requested for cgroup %s. Rules added: %d", v1Str, len(requestedAndCurrentRules)-len(r.Devices))
+	resourcesToSet.Devices = requestedAndCurrentRules
+
+	err = RunWithChroot(HostCgroupBasePath, func() error {
+		err := v.Manager.Set(&resourcesToSet)
+		return err
+	})
+
+	return logAndReturnErrorWithSprintfIfNotNil(err, errApplyingOtherRules, err)
+}
+
+func (v *v1Manager) GetCgroupVersion() string {
+	return v1Str
+}
+
+// addCurrentRules gets a slice of rules as a parameter and returns a new slice that contains all given rules
+// and all of the rules that are currently set. This way rules that are already defined won't be deleted by this
+// current request. Every old rule that is part of the new request will be overridden.
+//
+// For example, if the following rules are defined:
+// 1) {Minor: 111, Major: 111, Allow: true}
+// 2) {Minor: 222, Major: 222, Allow: true}
+//
+// And we get a request to enable the following rule: {Minor: 222, Major: 222, Allow: false}
+// Than we expect rule (1) to stay unchanged. Not adding it to our request will automatically delete it.
+func (v *v1Manager) addCurrentRules(deviceRules []*devices.Rule) ([]*devices.Rule, error) {
```

Comment:
I think current implementation is hard to cover with unit tests as it uses real filesystem paths and cgroup files, etc. That also applies to `v2Manager`.  IMHO all the external dependencies are better decoupled by introducing abstract interfaces that can be mocked for testing.

Reply:
There is an abstract interface defined in `pkg/virt-handler/cgroup/cgroup.go`:
```
type Manager interface {
	Set(r *configs.Resources) error
	runc_cgroups.Manager

	// GetBasePathToHostSubsystem returns the path to the specified subsystem
	// from the host's viewpoint.
	GetBasePathToHostSubsystem(subsystem string) (string, error)

	// GetCgroupVersion returns the current cgroup version (i.e. v1 or v2)
	GetCgroupVersion() string
}
```

This interface embeds runc's interface.
It can be used for testing. See my other comment about tests in general